tests/unit/Authentication.spec.js

```diff
@@ -0,0 +1,61 @@
+import { createLocalVue, shallowMount } from '@vue/test-utils';
+import Vuetify from 'vuetify';
+import RestClient from '@/rest';
+import Authentication from '@/components/Authentication/Authentication.vue';
+import Register from '@/components/Authentication/Register.vue';
+import Oauth from '@/components/Authentication/OAuth.vue';
+
+const localVue = createLocalVue();
+localVue.use(Vuetify);
+
+describe('Authentication', () => {
+  it('does not render children when toggled off', () => {
+    const wrapper = shallowMount(Authentication, {
+      localVue,
+      propsData: {
+        register: false,
+        oauth: false,
+        forgetPasswordLink: 'https://example.com/',
+      },
+      provide: { girderRest: new RestClient() },
+    });
+    expect(wrapper.contains(Register)).toBe(false);
+    expect(wrapper.contains(Oauth)).toBe(false);
+  });
+
+  it('renders registration when toggled on', () => {
+    const wrapper = shallowMount(Authentication, {
+      localVue,
+      propsData: {
+        register: true,
+        oauth: false,
+        forgetPasswordLink: 'https://example.com/',
+      },
+      provide: { girderRest: new RestClient() },
+    });
+    expect(wrapper.contains(Register)).toBe(true);
+  });
+
+  it('fetches Oauth Providers if Oauth is enabled', (done) => {
+    const get = () => Promise.resolve({
```

Comment:
We've created a nice pattern for mocking REST interaction that can be seen in uploadUtil.spec.js (among others).

We use [axios-mock-adapter](https://github.com/girder/girder_web_components/blob/fc44f25ce350bdf89647b310c5c77564c5ae1a02/tests/unit/uploadUtil.spec.js#L8) to mock the XHRs coming out of the RestClient, [like this example](https://github.com/girder/girder_web_components/blob/fc44f25ce350bdf89647b310c5c77564c5ae1a02/tests/unit/uploadUtil.spec.js#L55). For this, it'd look like:

```javascript
mock.onGet(/oauth\/provider/).replyOnce(200, [{
  id: 'foo',
  name: 'bar',
  url: 'https://example.com/',
}]);
```

The nice thing about using this way is that it makes it quite easy to maintain the test mocking even as components become more complex and make more requests, and `axios-mock-adapter` makes it very straightforward to mock rejected XHRs of all types (network failure, HTTP error responses, or timeouts).

Reply:
There are some things I likely don't understand about what this mocker is doing.

With my solution that directly uses `Promise.resolve()`, the tests pass with a single call to `await nextTick()`

If I replace that with `mock.onGet()`, I need 3 or more calls.  I checked and there's a 6-10 millisecond delay between the call and response for both.  This also shouldn't make a difference to the vue instance.

I tried experimenting with returning a promise from mock.onGet, but this still required exactly 3 calls to nextTick.

Reply:
That's weird. I hadn't tried this on an actual Vue component, so I haven't run into this myself. I will take a look at it today and see if I can sort it out, if so I'll just append a commit onto this branch and we can merge, otherwise we can use your previous solution for now.

Reply:
It might have to do with `mounted` being an async function, which interferes somehow with how vue-test-utils tries to make components synchronous.  There is a lot of traffic in the Vue issue tracker about this kind of thing.  The [vue-test-utils docs](https://vue-test-utils.vuejs.org/guides/#testing-asynchronous-behavior) mention using [flush-promises](https://github.com/kentor/flush-promises).  That seems to avoid the need for multiple `await`'s in this test.

As a side note on the oauth list, I am a fan of using [vue-async-computed](https://github.com/foxbenjaminfox/vue-async-computed) for async properties.  It might be too much to bring in for this one case, but I suspect it is a pattern we will need again.

Reply:
Agreed

Reply:
Test has been commented out with a reference to this thread.